subsys/net/lib/tftp/tftp_client.c

```diff
@@ -5,292 +5,247 @@
- * give up.
- * Return Value: This function will return the "opcode" received from
- * the remote server, i.e.
- * -> ERROR_OPCODE: If the remote server responded with "Error" or if
- *    the client was unable to send / receive anything from the server.
- * -> DATA_OPCODE:  If the remote server responded with "Data".
- * -> ACK_OPCODE:   If the remote server responded with "Ack".
- */
-static int tftp_send_request(int sock, uint8_t request,
-			     const char *remote_file, const char *mode)
-{
-	uint8_t    retx_cnt = 0;
-	uint32_t   req_size;
+		LOG_DBG("Sending TFTP request %d file %s", request,
+			log_strdup(remote_file));
 
-	/* Create TFTP Request. */
-	req_size = make_request(remote_file, mode, request);
+		/* Send the request to the server */
+		ret = sendto(sock, tftpc_buffer, req_size, 0, server_addr,
+			     ADDRLEN(server_addr));
+		if (ret < 0) {
+			break;
+		}
 
-send_req:
+		/* Poll for the response */
+		struct pollfd fds = {
+			.fd     = sock,
+			.events = ZSOCK_POLLIN,
+		};
+
+		ret = poll(&fds, 1, CONFIG_TFTPC_REQUEST_TIMEOUT);
+		if (ret <= 0) {
+			LOG_DBG("Failed to get data from the TFTP Server"
+				", req. no. %d", tx_count);
+			continue;
+		}
 
-	/* Send this request to the server. */
-	if (send(sock, tftpc_buffer, req_size, 0)) {
-		if (tftpc_recv_data(sock)) {
-			return sys_get_be16(tftpc_buffer);
+		/* Receive data from the TFTP Server. */
+		struct sockaddr from_addr;
+		socklen_t from_addr_len = sizeof(from_addr);
+
+		ret = recvfrom(sock, tftpc_buffer, TFTPC_MAX_BUF_SIZE, 0,
+				&from_addr, &from_addr_len);
+		if (ret < TFTP_HEADER_SIZE) {
+			req_size = make_request(tftpc_buffer, request,
+						remote_file, mode);
+			continue;
 		}
-	}
 
-	/* No of times we had to re-tx this "request". */
-	retx_cnt++;
+		/* Limit communication to the specific address:port */
+		connect(sock, &from_addr, from_addr_len);
 
-	/* Log this out. */
-	LOG_DBG("Unable to get data from the TFTP Server.");
-	LOG_DBG("no_of_retransmists = %d", retx_cnt);
+		break;
 
-	/* Are we retransmitting? */
-	if (retx_cnt < TFTP_REQ_RETX) {
-		LOG_DBG("Are we re-transmitting: YES");
-		goto send_req;
-	}
+	} while (tx_count <= TFTP_REQ_RETX);
 
-	LOG_DBG("Are we re-transmitting: NO");
-	return ERROR_OPCODE;
+	return ret;
 }
 
-/* Name: tftp_get
- * Description: This function gets "file" from the remote server.
- */
-int tftp_get(struct sockaddr *server, struct tftpc *client,
+int tftp_get(struct sockaddr *server_addr, struct tftpc *client,
 	     const char *remote_file, const char *mode)
 {
-
-	int32_t   stat     = TFTPC_UNKNOWN_FAILURE;
-	uint8_t    retx_cnt = 0;
-	int32_t   sock;
-
-	/* Re-init the global "block number" variable. */
-	tftpc_block_no = 1;
-	tftpc_index    = 0;
-
-	/* Create Socket for TFTP (IPv4 / UDP) */
-	sock = socket(server->sa_family, SOCK_DGRAM, IPPROTO_UDP);
+	int sock;
+	uint32_t tftpc_block_no = 1;
+	uint32_t tftpc_index = 0;
+	int tx_count = 0;
+	struct tftphdr_ack ackhdr = {htons(ACK_OPCODE), htons(1)};
```

Comment:
better would be set the values using the `.field = value,` notation, then there are no errors if the fields are reorganized in the struct